src/wp-admin/includes/image.php

```diff
@@ -1143,3 +1268,58 @@ function _copy_image_file( $attachment_id ) {
 
 	return $dst_file;
 }
+
+/**
+ * Returns an array with the list of valid mime types that a specific mime type should be converted into.
+ * For example an `image/jpeg` should be converted into an `image/jpeg` and `image/webp`. The first type
+ * is considered the primary output type for this image.
+ *
+ * @since 6.0.0
+ *
+ * @param $attachment_id int The attachment ID.
+ * @return array<string, array<string>> An array of valid mime types, where the key is the mime type and the value is the extension type.
+ */
+function wp_upload_image_mime_transforms( $attachment_id ) {
+	$image_mime_transforms = array(
+		'image/jpeg'      => array( 'image/jpeg', 'image/webp' ),
+		'image/webp'      => array( 'image/webp', 'image/jpeg' ),
+		'application/pdf' => array( 'image/jpeg' ),
+	);
+
+	/**
+	 * Filter to the output mime types for a given input mime type.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array $image_mime_transforms A map with the valid mime transforms.
+	 * @param int   $attachment_id The ID of the attachment where the hook was dispatched.
+	 */
+	return (array) apply_filters( 'wp_upload_image_mime_transforms', $image_mime_transforms, $attachment_id );
+}
+
+/**
+ * Extract the primary and additional mime output types for an image from the $image_mime_transforms.
+ *
+ * @since 6.0.0
+ *
+ * @param $image_mime_transforms array<string, array<string>> An array of valid mime types, where the key is the mime type and the value is the extension type.
+ * @param $attachment_id int The attachment ID.
+ * @param $mime_type string The mime type of the image.
+ */
+function _wp_get_primary_and_additional_mime_types( $image_mime_transforms, $attachment_id, $mime_type ) {
+	$mime_type         = get_post_mime_type( $attachment_id );
+	$output_mime_types = isset( $image_mime_transforms[ $mime_type ] ) ? $image_mime_transforms[ $mime_type ] : array( $mime_type );
+
+	// Use original mime type as primary mime type, or alternatively the first one.
+	$primary_mime_type_key = array_search( $mime_type, $output_mime_types, true );
+	if ( false === $primary_mime_type_key ) {
+		$primary_mime_type_key = 0;
+	}
+	// Split output mime types into primary mime type and additional mime types.
+	$additional_mime_types     = $output_mime_types;
+	list( $primary_mime_type ) = array_splice( $additional_mime_types, $primary_mime_type_key, 1 );
+	return array(
+		'primary_mime_type'     => $primary_mime_type,
+		'additional_mime_types' => $additional_mime_types,
+	);
```

Comment:
We should probably change this to just return an indexed array with these two values so that it can easily be used, with the PHP `list` construct. Otherwise you end up writing at least three lines of code just to call this function.

Especially since this is a private function, it's perfectly fine to return the data like this. Similar return values also exist in some other core functions.

If you update this, remember to also update the `@return` docs accordingly.

Reply:
👍🏼  Good suggestion, will update

Reply:
updated